### PR TITLE
Reduce dependnecies of ra_vfs

### PR DESCRIPTION
**Pull request by matklad**
In preparation for moving `ra_vfs` to a separate repo with extensive cross-platform CI, remove dependency on `ra_thread_workder` and `ra_arena`.

### CODE REVIEW

File: crates/ra_vfs/src/io.rs [hunk 1]
```diff
@@ -46,7 +47,40 @@ enum ChangeKind {
 
 const WATCHER_DELAY: Duration = Duration::from_millis(250);
 
-pub(crate) type Worker = thread_worker::Worker<Task, TaskResult>;
+// Like thread::JoinHandle, but joins the thread on drop.
+//
+// This is useful because it guarantees the absence of run-away threads, even if
+// code panics. This is important, because we might seem panics in the test and
```
Comment (pnkfelix) [inline crates/ra_vfs/src/io.rs]: typo: s/seem/see/

**Comment by matklad** [comment]
The repo is optimistically set up here: https://github.com/rust-analyzer/ra_vfs

File: crates/ra_vfs/src/io.rs [hunk 2]
```diff
@@ -46,7 +47,40 @@ enum ChangeKind {
...
+    fn drop(&mut self) {
+        let res = self.0.take().unwrap().join();
+        if !thread::panicking() {
+            res.unwrap();
+        }
+    }
+}
+
+pub(crate) struct Worker {
+    // XXX: it's important to drop `sender` before `_thread` to avoid deadlock.
```
Comment (pnkfelix) [inline crates/ra_vfs/src/io.rs]: you might consider including the detail that the desired drop order is accomplished here via the field declaration order. At least, I don't know if its something that is considered common knowledge in the Rust dev community. (The original `Worker` docs [here](https://github.com/rust-analyzer/rust-analyzer/blob/bd2b4ef2d1c631215523f79a8138cfa8cbd3f70e/crates/thread_worker/src/lib.rs#L42) did include that detail.)

**Comment by pnkfelix** [comment]
r=me once commit nits are addressed.

**Comment by matklad** [comment]
bors r=pnkfelix

**Pull request merged**